cmd/snap-bootstrap/hybrid_users_import.go

```diff
@@ -0,0 +1,462 @@
+			return false
+		}
+
+		// only consider users that are in the groups that we're specifically
+		// importing
+		for _, tg := range targetGroups {
+			if strutil.ListContains(u.groups, tg) {
+				return true
+			}
+		}
+
+		return false
+	}
+}
+
+func groupFilter(users map[string]user) func(group) bool {
+	return func(g group) bool {
+		// always import the root group
+		if g.gid == 0 {
+			return true
+		}
+
+		// in addition to importing the users that are in the given groups, we also
+		// import the primary groups that those users are in. these will most likely
+		// all be unique, but there is a chance that users share a primary group.
+		for _, u := range users {
+			if u.gid == g.gid {
+				return true
+			}
+		}
+
+		return false
+	}
+}
+
+func mergeAndWriteLoginFiles(importRoot, referenceRoot string, targetGroups []string, outputDir string) error {
+	users, err := parseUsers(importRoot, userFilter(targetGroups))
+	if err != nil {
+		return err
+	}
+
+	groups, err := parseGroups(importRoot, groupFilter(users))
+	if err != nil {
+		return err
+	}
+
+	// update the passwd and shadow files to contain the lines for the users
+	// that we're importing
+	err = mergeAndWriteUserFiles(referenceRoot, outputDir, users)
+	if err != nil {
+		return err
+	}
+
+	// update the group and gshadow files to add our users to any existing
+	// groups, and add any new primary groups that we're importing.
+	if err := mergeAndWriteGroupFiles(referenceRoot, outputDir, users, groups); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func mergeAndWriteUserFiles(originalRoot string, outputDir string, users map[string]user) error {
+	// only import users that are either root or have a uid < 1000 and gid <
+	// 1000. the base shouldn't contain anything that doesn't fit this criteria,
+	// but it is best to be safe.
+	sourceUsers, err := parseUsers(originalRoot, func(u user) bool {
+		return (u.gid == 0 && u.uid == 0) || (u.uid < 1000 && u.gid < 1000)
+	})
+	if err != nil {
+		return err
+	}
+
+	var passwdBuffer bytes.Buffer
+	var shadowBuffer bytes.Buffer
+	for name, user := range sourceUsers {
+		if name != user.name {
+			return fmt.Errorf("internal error: user entry inconsistent with parsed data")
+		}
+
+		// if there is a conflict in the existing file, we will use the new
+		// entry instead of the old one. this is consistent with how we handle
+		// groups, so any conflicting users will always be fully replaced.
+		if _, ok := users[name]; ok {
+			continue
+		}
+
+		passwdBuffer.WriteString(user.passwdEntry + "\n")
+		if user.shadowEntry != "" {
+			shadowBuffer.WriteString(user.shadowEntry + "\n")
+		}
+	}
+
+	for _, user := range users {
+		// only allow importing users that are either root or have a uid < 1000
+		// and gid < 1000
+		if (user.uid != 0 || user.gid != 0) && (user.uid < 1000 || user.gid < 1000) {
+			return fmt.Errorf("internal error: cannot import system users: %s has uid %d and gid %d", user.name, user.uid, user.gid)
+		}
```

Comment:
We already filtered, imo it is a bit too much to have this internal check for something already done in a nearby function, maybe a comment should suffice.